bundle/v2/client.go

```diff
@@ -0,0 +1,583 @@
+func (c *Client) GetBundle(ctx context.Context, bundleLabel string) (string, error) {
+	log := c.Logger.WithValues("bundle", bundleLabel)
+	log.V(1).Info("Calling GetBundle RPC")
+
+	resp, err := c.rpcClient.GetBundle(ctx, connect.NewRequest(&bundlev2.GetBundleRequest{
+		PdpId:       c.PDPIdentifier,
+		BundleLabel: bundleLabel,
+	}))
+	if err != nil {
+		log.Error(err, "GetBundle RPC failed")
+		return "", err
+	}
+
+	base.LogResponsePayload(log, resp.Msg)
+
+	return c.getBundleFile(logr.NewContext(ctx, log), resp.Msg.BundleInfo)
+}
+
+func (c *Client) WatchBundle(ctx context.Context, bundleLabel string) (bundle.WatchHandle, error) {
+	log := c.Logger.WithValues("bundle", bundleLabel)
+	log.V(1).Info("Calling WatchBundle RPC")
+
+	stream := c.rpcClient.WatchBundle(ctx)
+	if err := stream.Send(nil); err != nil {
+		log.V(1).Error(err, "Failed to send request headers")
+		_ = stream.CloseRequest()
+		_ = stream.CloseResponse()
+		return nil, fmt.Errorf("failed to send request headers: %w", err)
+	}
+
+	wh := &bundle.WatchHandleImpl{
+		ServerEventsCh: make(chan bundle.ServerEvent, 1),
+		ClientEventsCh: make(chan bundle.ClientEvent, 1),
+		ErrorsCh:       make(chan error, 1),
+		BundleLabel:    bundleLabel,
+		Pool:           pool.New().WithContext(ctx).WithCancelOnError().WithFirstError(),
+	}
+
+	wh.Pool.Go(c.watchStreamRecv(stream, wh, log))
+	wh.Pool.Go(c.watchStreamSend(stream, wh, log))
+	go func() {
+		if err := wh.Wait(); err != nil {
+			switch {
+			case errors.Is(err, context.Canceled):
+				log.V(2).Info("Watch stream terminated: context cancelled")
+			case errors.Is(err, bundle.ErrBundleNotFound):
+				log.V(2).Info("Watch stream terminated: bundle not found")
+			case errors.As(err, &bundle.ReconnectError{}):
+				log.V(2).Info("Watch stream terminated: server requests reconnect")
+			default:
+				log.V(2).Error(err, "Watch stream terminated")
+			}
+			return
+		}
+		log.V(2).Info("Watch stream terminated")
+	}()
+
+	return wh, nil
+}
+
+type recvMsgV2 struct {
+	msg *bundlev2.WatchBundleResponse
+	err error
+}
+
+func (c *Client) watchStreamRecv(stream *connect.BidiStreamForClient[bundlev2.WatchBundleRequest, bundlev2.WatchBundleResponse], wh *bundle.WatchHandleImpl, logger logr.Logger) func(context.Context) error {
+	return func(ctx context.Context) (outErr error) {
+		log := logger.WithName("recv")
+		log.V(1).Info("Starting response handler")
+
+		recvChan := make(chan recvMsgV2, 1)
+		go func() {
+			defer func() {
+				close(recvChan)
+				if err := stream.CloseResponse(); err != nil && !errors.Is(err, context.Canceled) {
+					log.V(2).Error(err, "Error while closing response stream")
+				}
+			}()
+
+			for {
+				if err := ctx.Err(); err != nil {
+					return
+				}
+
+				log.V(3).Info("Waiting to receive message")
+				msg, err := stream.Receive()
+				log.V(3).Info("Received message")
+
+				recvChan <- recvMsgV2{msg: msg, err: err}
+				if err != nil {
+					if errors.Is(err, context.Canceled) {
```

Comment:
https://github.com/cerbos/cloud-api/pull/229
```suggestion
					if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
```